plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange.hs

```diff
@@ -0,0 +1,144 @@
+    -- 'positionMapping' should be applied to the input positions before using them
+    positions' <- maybeToExceptT "fail to apply position mapping to input positions" . MaybeT . pure $
+        traverse (fromCurrentPosition positionMapping) positions
+
+    ast <- maybeToExceptT "fail to get ast for current file" . MaybeT . pure $
+        -- in GHC 9, the 'FastString' in 'HieASTs' is replaced by a newtype wrapper around 'LexicalFastString'
+        -- so we use 'coerce' to make it work in both GHC 8 and 9
+        getAsts hieAst Map.!? (coerce . mkFastString . fromNormalizedFilePath) file
+
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    let selectionRanges = findSelectionRangesByPositions (astPathsLeafToRoot ast') positions'
+
+    -- 'positionMapping' should be applied to the output ranges before returning them
+    maybeToExceptT "fail to apply position mapping to output positions" . MaybeT . pure $
+         traverse (toCurrentSelectionRange positionMapping) selectionRanges
+
+-- | Likes 'toCurrentPosition', but works on 'SelectionRange'
+toCurrentSelectionRange :: PositionMapping -> SelectionRange -> Maybe SelectionRange
+toCurrentSelectionRange positionMapping SelectionRange{..} = do
+    newRange <- toCurrentRange positionMapping _range
+    pure $ SelectionRange {
+        _range = newRange,
+        _parent = _parent >>= toCurrentSelectionRange positionMapping
+    }
+
+-- | Build all paths from ast leaf to root
+astPathsLeafToRoot :: HieAST a -> [SelectionRange]
+astPathsLeafToRoot = mapMaybe (spansToSelectionRange . nubOrd) . go [[]]
+  where
+    go :: [[Span]] -> HieAST a -> [[Span]]
+    go acc (Node _ span [])       = fmap (span:) acc
+    go acc (Node _ span children) = concatMap (go (fmap (span:) acc)) children
+
+spansToSelectionRange :: [Span] -> Maybe SelectionRange
+spansToSelectionRange [] = Nothing
+spansToSelectionRange (span:spans) = Just $
+    SelectionRange {_range = realSrcSpanToRange span, _parent = spansToSelectionRange spans}
+
+{-|
+For each position, find the selection range that contains it, without taking each selection range's
+parent into account. These selection ranges are un-divisible, representing the leaf nodes in original AST, so they
+won't overlap.
+-}
+findSelectionRangesByPositions :: [SelectionRange] -- ^ all possible selection ranges
+                               -> [Position] -- ^ requested positions
+                               -> [SelectionRange]
+findSelectionRangesByPositions selectionRanges = fmap findByPosition
+    {-
+        Performance Tips:
+        Doing a linear search from the first selection range for each position is not optimal.
+        If it becomes too slow for a large file and many positions, you may optimize the implementation.
+        At least we don't need to search from the very beginning for each position, if the are sorted firstly.
+        Or maybe we could apply some techniques like binary search?
```

Comment:
Yes please, do consider at least benchmarking this in a follow-up PR. We do care about performance!  Unfortunately the benchmark suite doesn't currently support plugins (it's a benchmark suite for `ghcide`, not HLS). 

I wonder if binary search is really the solution here.  I suspect `HieAst` is isomorphic to a position indexed search tree, so the optimal approach would take advantage of this to find the solution  without building a list of all selection ranges.

Reply:
Oh, that's another optimization I didn't come up with. So if I get it correctly, there are two known optimizations now. (let n represents the number of selection ranges, m represents the number of positions)

+ For each position, we may treat HieAST as a position indexed tree to search it in O(log(n)).
+ For all positions, a searched position will narrow the search range for other positions. If we search the middle position first, then all positions before this position may only need to search in half of the tree (if both the tree and positions are balanced)

So the final complexity will be O(log(n) * log(m)).

If this looks right, I will update the comment to reflect these thoughts.

Reply:
Right, at the moment (I think) you build all the possible selection ranges in the document, and then filter them by position, when you could fuse that into one pass and only build the paths which contain the target position or something.

Reply:
Maybe it's worth noting that, as selection ranges have shared parents, building up the list is O(n), where n is the number of nodes in the original AST. `astPathsLeafToRoot` basically clone the tree and make every pointer reversed. And functions in `ASTPreProcess` need to traverse through the AST, it's also O(n). So maybe this is not the only performance bottleneck, and we will need furthur benchmarks to make a conclusion.

Reply:
Not that it matters much but I think laziness saves a 1/2 constant factor on average for building the root to leaf paths. I think this holds true for the preprocessing phase too even with the `reverse` `foldl'` but not certain.

So if the preprocessing phase is lazy enough, then I think if we needed to upgrade the algorithm applying it to only `astPathsLeafToRoot` might suffice.